app/src/main/java/com/duckduckgo/app/generalsettings/GeneralSettingsViewModel.kt

```diff
@@ -0,0 +1,152 @@
+            voiceSearchEnabled = voiceSearchAvailability.isVoiceSearchAvailable,
+        ),
+    )
+
+    fun viewState(): Flow<ViewState> = viewState.onStart {
+        viewModelScope.launch(dispatcherProvider.io()) {
+            val autoCompleteEnabled = settingsDataStore.autoCompleteSuggestionsEnabled
+            if (!autoCompleteEnabled) {
+                history.setHistoryUserEnabled(false)
+            }
+            withContext(dispatcherProvider.main()) {
+                viewState.emit(
+                    ViewState(
+                        autoCompleteSuggestionsEnabled = settingsDataStore.autoCompleteSuggestionsEnabled,
+                        autoCompleteRecentlyVisitedSitesSuggestionsUserEnabled = history.isHistoryUserEnabled(),
+                        storeHistoryEnabled = history.isHistoryFeatureAvailable(),
+                        showVoiceSearch = voiceSearchAvailability.isVoiceSearchSupported,
+                        voiceSearchEnabled = voiceSearchAvailability.isVoiceSearchAvailable,
+                    ),
+                )
+            }
+        }
+    }
+
+    fun onAutocompleteSettingChanged(enabled: Boolean) {
+        Timber.i("User changed autocomplete setting, is now enabled: $enabled")
+        viewModelScope.launch(dispatcherProvider.io()) {
+            settingsDataStore.autoCompleteSuggestionsEnabled = enabled
+            if (!enabled) {
+                viewModelScope.launch() {
+                    history.setHistoryUserEnabled(false)
+                }
+            }
+            if (enabled) {
+                pixel.fire(AUTOCOMPLETE_GENERAL_SETTINGS_TOGGLED_ON)
+            } else {
+                pixel.fire(AUTOCOMPLETE_GENERAL_SETTINGS_TOGGLED_OFF)
+            }
+            withContext(dispatcherProvider.main()) {
+                viewState.emit(
+                    currentViewState().copy(
+                        autoCompleteSuggestionsEnabled = enabled,
+                        autoCompleteRecentlyVisitedSitesSuggestionsUserEnabled = history.isHistoryUserEnabled(),
+                    ),
+                )
+            }
+        }
+    }
+
+    fun onAutocompleteRecentlyVisitedSitesSettingChanged(enabled: Boolean) {
+        Timber.i("User changed autocomplete recently visited sites setting, is now enabled: $enabled")
+        viewModelScope.launch(dispatcherProvider.io()) {
+            history.setHistoryUserEnabled(enabled)
+            if (enabled) {
+                pixel.fire(AUTOCOMPLETE_RECENT_SITES_GENERAL_SETTINGS_TOGGLED_ON)
+            } else {
+                pixel.fire(AUTOCOMPLETE_RECENT_SITES_GENERAL_SETTINGS_TOGGLED_OFF)
+            }
+            withContext(dispatcherProvider.main()) {
+                viewState.emit(currentViewState().copy(autoCompleteRecentlyVisitedSitesSuggestionsUserEnabled = enabled))
+            }
+        }
+    }
+
+    fun onVoiceSearchChanged(checked: Boolean) {
+        voiceSearchRepository.setVoiceSearchUserEnabled(checked)
```

Comment:
should you wrap this also in a coroutine?

Reply:
This is copied from `AccessibilitySettingsViewModel`

Reply:
I’ve moved everything in `onVoiceSearchChanged` into a coroutine, here and in `AccessibilitySettingsViewModel`